proposed/0006-unified-structured-data-imporvement.md

```diff
@@ -0,0 +1,177 @@
+ 
+**step 8:**
+Client-B tries to update:
+```
+StructuredData {
+    other_fields: Value-33,
+    owner_keys: B-PublicKey,
+    previous_owner_keys: NONE,
+    signatures: Signed by B-PrivateKey,
+}
+```
+ 
+**step 9:**
+Logic: Since `previous_owner_keys` in incoming data contains nothing, use `key` = `(sha512(owner_keys), Name)` for look-up.
+Vault will check if new data is sent by valid owner. Use `owner_keys` from the stored data in step 5 to verify signature of data in step 8. If valid replace previous data and store new one:
+```
+Key = (sha512(B-PublicKey), Name)
+Value = StructuredData {
+    other_fields: Value-33,
+    owner_keys: B-PublicKey,
+    previous_owner_keys: NONE,
+    signatures: Signed by B-PrivateKey,
+}
+```
+## Change in the Definition of GETs
+The above shows the cases for PUTs, POSTs and DELETEs. There will be a change in the signature of GETs to achieve the above no-collision scheme. GETs for StructuredData need to be changed from:
+```
+GET(Name, DataRequest::StructuredData(u64))
+```
+to
+```
+GET(Name, DataRequest::StructuredData(SHA512(PublicKeys)))
+```
+For ordinary cases where `data-PUTers` themselves are fetching thier data (majority of clients) this does not mean much as the client engine will inject this silently as it already has the knowledge of user public-keys. The only time it does not have the knowledge is during the fetching of the session packet. The solution for this is easy as login packet has a dedicated maidasafe-type-tag and vaults can recognise it and make special allowance for it to be stored with `Key = Name` instead of `Key = (sha512(PublicKeys), Name)` as in above.
+
+#Drawbacks
+For other cases where the retrieval is done for the data PUT by others, the GETer must know the PublicKey of the original PUTer. An example of this will be the fetching of Dns Packets by a safe-url-parser (could be a browser). This may be seen as an advantage instead of disadvantage though as explained thus. Since there is no retriction for taking a Dns name in the SAFE-network, a user can very easily obtain well known Dns names like `pepsico.com` preventing the legit organisation to take it later. However with the schema in thi rfc, the legit organisation can easily obtain the same Dns name for it because it will have got the uniqueness in the vaults due to `sha512` of its PublicKey. This PublicKey will be the one which is distributed to all browsers and thus safe-url-parser will land on the legitimate page instead of the pretender's.
+
+Since Dns Packets also have special type-tags, the vaults can recognise them. There could be ordinary ones (with TAG = X) where the storage is `Key = Name` for which there would be collisions and special ones (with TAG = Y) where the storage would be `Key = (sha512(PublicKey), Name)`. The safe-browser would land on the correct destination depending on whether TAG = X or TAG = Y was demanded. This is beyond the scope of this Rfc and can be discussed in a separate one.
+
```

Comment:
It has to be based on type-tag at some level. Consider this: You are putting an SD onto the vaults - the vaults don't take tag into consideration. Every `PUT`ter has a valid MaidSafe Account and thus valid PublicKeys. So PublicKey will always be known for a `PUT` and vaults would never have to give a default value. However the public-key might  not be known for `GET`s for that particular SD - eg., fetching of a page by a browser - it wouldn't know the PublicKey of the user who `PUT` the SD Record. In that case if you default it to something, the algorithm's not going to work at all because you will never end up actually finding it.